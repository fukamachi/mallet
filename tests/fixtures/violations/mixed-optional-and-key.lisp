;; Test fixtures for mixed &optional and &key rule

;; Bad: defun mixing &optional and &key
(defun process-data (input &optional format &key validate)
  (list input format validate))

;; Good: defun with only &key
(defun process-data-good (input &key format validate)
  (list input format validate))

;; Bad: lambda mixing &optional and &key
(lambda (x &optional y &key z)
  (list x y z))

;; Good: lambda with only &optional
(lambda (x &optional y)
  (list x y))

;; Bad: defmethod mixing &optional and &key
(defmethod compute ((obj string) &optional precision &key units)
  (list obj precision units))

;; Good: defmethod with only &key
(defmethod compute-good ((obj string) &key precision units)
  (list obj precision units))

;; Bad: defmacro mixing &optional and &key
(defmacro with-context (name &optional default &key scope)
  `(let ((,name (or ,default (get-default ,scope))))
     ,@body))

;; Good: defmacro with only &key
(defmacro with-context-good (name &key default scope)
  `(let ((,name (or ,default (get-default ,scope))))
     ,@body))

;; Bad: flet mixing &optional and &key
(flet ((helper (a &optional b &key c)
         (list a b c)))
  (helper 1))

;; Good: flet with only &key
(flet ((helper-good (a &key b c)
         (list a b c)))
  (helper-good 1))
