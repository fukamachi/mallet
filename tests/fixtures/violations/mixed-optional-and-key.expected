# Expected violations for mixed-optional-and-key.lisp
# Format: line:column rule-name severity

4:20 mixed-optional-and-key warning
12:8 mixed-optional-and-key warning
20:19 mixed-optional-and-key warning
28:23 mixed-optional-and-key warning
38:15 mixed-optional-and-key warning
